File: eCommerce/eCommerce.Services/Database/Reservation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using eCommerce.Model;

namespace eCommerce.Services.Database
{
    public class Reservation
    {
        [Key]
        public int Id { get; set; }
        
        public int UserId { get; set; }
        
        [ForeignKey("UserId")]
        public User User { get; set; } = null!;
        
        public int RestaurantId { get; set; }
        
        [ForeignKey("RestaurantId")]
        public Restaurant Restaurant { get; set; } = null!;
        
        public int? TableId { get; set; }
        
        [ForeignKey("TableId")]
        public Table? Table { get; set; }
        
        public DateTime ReservationDate { get; set; }
        
        public TimeSpan ReservationTime { get; set; }
        
        public TimeSpan Duration { get; set; } = TimeSpan.FromHours(2); // Default 2 hours
        
        public int NumberOfGuests { get; set; }
        
        /// <summary>
        /// Gets the current state of the reservation.
        /// State can only be changed through the state machine methods (Edit, Confirm, Cancel, Complete).
        /// </summary>
        [Required]
        [Column(TypeName = "int")]
        public ReservationState State { get; private set; } = ReservationState.Requested;
        
        [MaxLength(500)]
        public string? SpecialRequests { get; set; }
        
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        
        public DateTime? ConfirmedAt { get; set; }
        
        public DateTime? CancelledAt { get; set; }
        
        [MaxLength(500)]
        public string? CancellationReason { get; set; }
        
        // Navigation properties
        public ICollection<ReservationHistory> History { get; set; } = new List<ReservationHistory>();
        public Review? Review { get; set; }

        /// <summary>
        /// Computed property that returns the start time of the reservation
        /// (combines ReservationDate and ReservationTime).
        /// </summary>
        [NotMapped]
        public DateTime StartTime => ReservationDate.Date.Add(ReservationTime);

        /// <summary>
        /// Computed property that returns the end time of the reservation
        /// (StartTime + Duration).
        /// </summary>
        [NotMapped]
        public DateTime EndTime => StartTime.Add(Duration);

        /// <summary>
        /// Allows editing reservation details. Only allowed when State is Requested.
        /// This method allows modification of reservation properties such as ReservationDate,
        /// ReservationTime, Duration, NumberOfGuests, and SpecialRequests.
        /// </summary>
        /// <param name="reservationDate">New reservation date.</param>
        /// <param name="reservationTime">New reservation time.</param>
        /// <param name="duration">New duration.</param>
        /// <param name="numberOfGuests">New number of guests.</param>
        /// <param name="specialRequests">New special requests (optional).</param>
        /// <exception cref="InvalidOperationException">Thrown when State is not Requested.</exception>
        public void Edit(DateTime reservationDate, TimeSpan reservationTime, TimeSpan duration, int numberOfGuests, string? specialRequests = null)
        {
            if (State != ReservationState.Requested)
            {
                throw new InvalidOperationException($"Cannot edit reservation. Editing is only allowed when State is {ReservationState.Requested}, but current State is {State}.");
            }

            ReservationDate = reservationDate;
            ReservationTime = reservationTime;
            Duration = duration;
            NumberOfGuests = numberOfGuests;
            SpecialRequests = specialRequests;
        }

        /// <summary>
        /// Confirms the reservation. Only allowed when State is Requested.
        /// Sets the ConfirmedAt timestamp to the current UTC time.
        /// </summary>
        /// <exception cref="InvalidOperationException">Thrown when State is not Requested.</exception>
        public void Confirm()
        {
            if (State != ReservationState.Requested)
            {
                throw new InvalidOperationException($"Cannot confirm reservation. Confirmation is only allowed when State is {ReservationState.Requested}, but current State is {State}.");
            }

            State = ReservationState.Confirmed;
            ConfirmedAt = DateTime.UtcNow;
        }

        /// <summary>
        /// Cancels the reservation. Allowed when State is Requested or Confirmed.
        /// Sets the CancelledAt timestamp to the current UTC time and optionally records a cancellation reason.
        /// </summary>
        /// <param name="reason">Optional cancellation reason.</param>
        /// <exception cref="InvalidOperationException">Thrown when State is not Requested or Confirmed.</exception>
        public void Cancel(string? reason = null)
        {
            if (State != ReservationState.Requested && State != ReservationState.Confirmed)
            {
                throw new InvalidOperationException($"Cannot cancel reservation. Cancellation is only allowed when State is {ReservationState.Requested} or {ReservationState.Confirmed}, but current State is {State}.");
            }

            State = ReservationState.Cancelled;
            CancelledAt = DateTime.UtcNow;
            CancellationReason = reason;
        }

        /// <summary>
        /// Marks the reservation as completed. Only allowed when State is Confirmed.
        /// This indicates that the dining experience has finished.
        /// </summary>
        /// <exception cref="InvalidOperationException">Thrown when State is not Confirmed.</exception>
        public void Complete()
        {
            if (State != ReservationState.Confirmed)
            {
                throw new InvalidOperationException($"Cannot complete reservation. Completion is only allowed when State is {ReservationState.Confirmed}, but current State is {State}.");
            }

            State = ReservationState.Completed;
        }

        /// <summary>
        /// Marks the reservation as expired. Only allowed when State is Requested.
        /// This is automatically called when a Requested reservation's EndTime has passed
        /// without being confirmed.
        /// </summary>
        /// <exception cref="InvalidOperationException">Thrown when State is not Requested.</exception>
        public void Expire()
        {
            if (State != ReservationState.Requested)
            {
                throw new InvalidOperationException($"Cannot expire reservation. Expiration is only allowed when State is {ReservationState.Requested}, but current State is {State}.");
            }

            State = ReservationState.Expired;
        }
    }
}

File: eCommerce/UI/ecommerce_mobile/lib/model/reservation.dart
class Reservation {
  final int id;
  final int userId;
  final String userName;
  final int restaurantId;
  final String restaurantName;
  final int tableId;
  final String tableNumber;
  final DateTime reservationDate;
  final String reservationTime; // Format: "HH:mm:ss"
  final String duration; // Format: "HH:mm:ss"
  final int numberOfGuests;
  final String status; // "Requested", "Confirmed", "Completed", "Cancelled", "Expired"
  final String? specialRequests;
  final DateTime createdAt;
  final DateTime? confirmedAt;
  final DateTime? cancelledAt;
  final String? cancellationReason;

  Reservation({
    required this.id,
    required this.userId,
    required this.userName,
    required this.restaurantId,
    required this.restaurantName,
    required this.tableId,
    required this.tableNumber,
    required this.reservationDate,
    required this.reservationTime,
    required this.duration,
    required this.numberOfGuests,
    required this.status,
    this.specialRequests,
    required this.createdAt,
    this.confirmedAt,
    this.cancelledAt,
    this.cancellationReason,
  });

  factory Reservation.fromJson(Map<String, dynamic> json) {
    return Reservation(
      id: json['id'] ?? 0,
      userId: json['userId'] ?? 0,
      userName: json['userName'] ?? '',
      restaurantId: json['restaurantId'] ?? 0,
      restaurantName: json['restaurantName'] ?? '',
      tableId: json['tableId'] ?? 0,
      tableNumber: json['tableNumber'] ?? '',
      reservationDate: json['reservationDate'] != null
          ? DateTime.parse(json['reservationDate'])
          : DateTime.now(),
      reservationTime: json['reservationTime'] ?? '00:00:00',
      duration: json['duration'] ?? '02:00:00',
      numberOfGuests: json['numberOfGuests'] ?? 0,
      status: json['status'] ?? 'Requested',
      specialRequests: json['specialRequests'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      confirmedAt: json['confirmedAt'] != null
          ? DateTime.parse(json['confirmedAt'])
          : null,
      cancelledAt: json['cancelledAt'] != null
          ? DateTime.parse(json['cancelledAt'])
          : null,
      cancellationReason: json['cancellationReason'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'userName': userName,
      'restaurantId': restaurantId,
      'restaurantName': restaurantName,
      'tableId': tableId,
      'tableNumber': tableNumber,
      'reservationDate': reservationDate.toIso8601String(),
      'reservationTime': reservationTime,
      'duration': duration,
      'numberOfGuests': numberOfGuests,
      'status': status,
      'specialRequests': specialRequests,
      'createdAt': createdAt.toIso8601String(),
      'confirmedAt': confirmedAt?.toIso8601String(),
      'cancelledAt': cancelledAt?.toIso8601String(),
      'cancellationReason': cancellationReason,
    };
  }

  // Helper method to calculate EndTime
  DateTime get endTime {
    final timeParts = reservationTime.split(':');
    final hours = int.parse(timeParts[0]);
    final minutes = int.parse(timeParts[1]);
    
    final durationParts = duration.split(':');
    final durationHours = int.parse(durationParts[0]);
    final durationMinutes = int.parse(durationParts[1]);
    
    return reservationDate.add(Duration(
      hours: hours + durationHours,
      minutes: minutes + durationMinutes,
    ));
  }

  // Helper method to check if reservation is in the past
  bool get isPast {
    return endTime.isBefore(DateTime.now());
  }
}
